### PR TITLE
core: remove __early_bss

### PR DESCRIPTION
**Pull request by jforissier**
Initialize the .bss section early from assembler before entering C
code. As a result, the __early_bss qualifier is not needed anymore.
Remove it, as well as the related symbols (__early_bss_start and
__early_bss_end).

Tested on QEMU and HiKey (32 and 64 bits) with and without pager.

Signed-off-by: Jerome Forissier <jerome.forissier@linaro.org>

### CODE REVIEW

File: core/arch/arm/kernel/generic_entry_a32.S [hunk 1]
```diff
@@ -304,6 +304,14 @@ copy_init:
 	bgt	copy_init
 #endif
 
+	/* Clear .bss */
+	ldr	r0, =__bss_start
+	ldr	r1, =__bss_end
+	mov	r2, #0
+clear_bss:
+	strb	r2, [r0], #1
```
Comment (etienne-lms) [inline core/arch/arm/kernel/generic_entry_a32.S]: minor: could speed up a bit through `str r2, [r0], #4`. **kern.ld.S** already include the required alignment for the `__bss_start` and `__bss_end`. From few builds, it seems early bss was below 1kByte, while bss is below 100kB.
Same could apply to **generic_entry_a64.S**.

Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_entry_a32.S]: This function could also be optimized by using the `stm` instruction, but I think it's OK to take such optimizations later.

**Comment by jenswi-linaro** [comment]
I've discussed this change with @etienne-lms and I'm starting to doubt that this is the right thing to do, especially for v7 systems.

**Comment by jforissier** [comment]
@jenswi-linaro how would that be bad? Why for v7 systems especially? Are you concerned by the time overhead of clearing memory with cache turned off? I can probably instrument the code and provide numbers if needed. Are you thinking about other issues?
From the code maintenance standpoint, I think this change is good. I mean when you're writing C code you normally should be able to assume that the runtime environment has been set up properly by the time your code is invoked, and that includes zero initialization of globals I would think.

**Comment by etienne-lms** [comment]
> From the code maintenance standpoint, I think this change is good

From maintenance view (as from development view), I think it's better that those pieces of code acting in the early inits of the system should be aware of it and take such special care. Requiring such code to explicitly use `__early_bss` when relying on bss during very early inits can be good as it explicitly state that this code is used in specific conditions, as those piece of code needing to take care whether mmu is on or off.

**Comment by etienne-lms** [comment]
About overhead, what when there comes a piece of code that every platform need for a end-product feature (I mean not for a debugging usecase) and that require a 1MByte of bss data ? Such code would rather expects BSS is inited with caches on.

**Comment by jforissier** [comment]
Not all early init functions (those that need to check `cpu_mmu_enabled()`) would need variables in the "early bss" anyway so I don't see the `__early_bss` marker as a reliable indicator of whether one needs to be careful of the PA vs VA issue. It's not like it would be a annotation on the function for instance.

Regarding speed I will definitely measure it. In any case we don't have platforms with megabytes of .bss yet... (although we don't know what downstream OP-TEE users may be doing, indded).

Finally, if we decide to keep `__early_bss` I'd like to see it documented, that is: why it's there and when to use it (in `compiler.h` probably).

**Comment by jforissier** [comment]
I did some measurement using CNTPCT and CNTFRQ on HiKey. I measured the time it takes to clear .bss in assembler before MMU and caches are turned on.

- 3063 us (3 ms) in 32-bit mode (.bss size: 65328 bytes)
- 629 us in 32-bit mode with pager (.bss size: 11656 bytes) 
- 78 us in 64-bit mode (.bss size: 73368 bytes)

Then I instrumented the code in current master, and measured the `memset()` that clears .bss with MMU and caches turned on:

- 216 us in 32-bit mode (.bss size: 65016 bytes)
- 24 us in 32-bit mode with pager (.bss size: 11328 bytes)
- 182 us in 64-bit mode (.bss size: 72800 bytes)

**Comment by MrVan** [comment]
@jforissier Will this patch merged to 2.4.0? Since in core/arm/arm/plat-imx/psci/psci.c, there still a variable, use `__data` attribute. I would like to fix that. If this patch not merged, i will use __early_bss. If this patch merged, i'll drop the `__data`  attribute.

**Comment by jforissier** [comment]
@MrVan well, since it looks like this patch is a bit controversial, I don't think it will make it into 2.4.0. Let's keep this PR open for further discussions -- there's no hurry.
In the mean time, yes, please use __early_bss for now. Thanks!

**Comment by jforissier** [comment]
Updated. The arm32 loop now clears two 32-bit words at a time (`stmia`) instead of one, which makes it a bit faster (see commit message). Rebased on top of master.

File: core/arch/arm/kernel/generic_entry_a32.S [hunk 2]
```diff
@@ -316,6 +316,15 @@ copy_init:
 	bgt	copy_init
 #endif
 
+	/* Clear .bss */
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_entry_a32.S]:
```
/*
 * Clear .bss, this code obviously depends on the linker keeping
 * start/end of .bss at least 8 byte aligned.
 */
```

Comment (jforissier) [inline core/arch/arm/kernel/generic_entry_a32.S]: Yep, I'll update.

**Pull request merged**